test/unit/controllers/declaration/ItemTypeControllerSpec.scala

```diff
@@ -0,0 +1,245 @@
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "form is incorrect during saving" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val incorrectForm = Seq(
+          ("combinedNomenclatureCode", ""),
+          ("taricAdditionalCode[]", "!@#$$%"),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", ""),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", ""),
+          saveAndContinueActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(incorrectForm: _*))
+
+        status(result) mustBe BAD_REQUEST
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+    }
+
+    "return 303 (SEE_OTHER)" when {
+
+      "item doesn't exists in the cache" in {
+
+        withNewCaching(aDeclaration())
+
+        val result = controller.displayPage(itemId)(getRequest())
+
+        status(result) mustBe SEE_OTHER
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "correct item is added and declaration model exist in the cache" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val correctForm = Seq(
+          ("combinedNomenclatureCode", "code"),
+          ("taricAdditionalCode[]", ""),
+          ("nationalAdditionalCode[]", ""),
+          ("descriptionOfGoods", "description"),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", "1234"),
+          addActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(correctForm: _*))
+
+        status(result) mustBe OK
+        verify(mockItemTypePage, times(1)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "correct item is added during saving" in {
+
+        withNewCaching(aDeclaration(withItem(anItem(withItemId(itemId)))))
+
+        val correctForm = Seq(
+          ("combinedNomenclatureCode", "code"),
+          ("taricAdditionalCode[]", "1234"),
+          ("nationalAdditionalCode[]", "VATR"),
+          ("descriptionOfGoods", "description"),
+          ("cusCode", ""),
+          ("unDangerousGoodsCode", ""),
+          ("statisticalValue", "1234"),
+          saveAndContinueActionUrlEncoded
+        )
+
+        val result = controller.submitItemType(itemId)(postRequestAsFormUrlEncoded(correctForm: _*))
+
+        status(result) mustBe SEE_OTHER
+        verify(mockItemTypePage, times(0)).apply(any(), any(), any(), any(), any())(any(), any())
+      }
+
+      "item has been removed with model in cache" in {
```

Comment:
I think "piece of data has been removed and there is model in cache" explains better.